rust/datafusion/src/logical_plan/dfschema.rs

```diff
@@ -0,0 +1,415 @@
+        Self::new(
+            schema
+                .fields()
+                .iter()
+                .map(|f| DFField {
+                    field: f.clone(),
+                    qualifier: Some(qualifier.to_owned()),
+                })
+                .collect(),
+        )
+    }
+
+    /// Combine two schemas
+    pub fn join(&self, schema: &DFSchema) -> Result<Self> {
+        let mut fields = self.fields.clone();
+        fields.extend_from_slice(schema.fields().as_slice());
+        Self::new(fields)
+    }
+
+    /// Get a list of fields
+    pub fn fields(&self) -> &Vec<DFField> {
+        &self.fields
+    }
+
+    /// Returns an immutable reference of a specific `Field` instance selected using an
+    /// offset within the internal `fields` vector
+    pub fn field(&self, i: usize) -> &DFField {
+        &self.fields[i]
+    }
+
+    /// Find the index of the column with the given name
+    pub fn index_of(&self, name: &str) -> Result<usize> {
+        for i in 0..self.fields.len() {
+            if self.fields[i].name() == name {
+                return Ok(i);
+            }
+        }
+        Err(DataFusionError::Plan(format!("No field named '{}'", name)))
+    }
+
+    /// Find the field with the given name
+    pub fn field_with_name(
+        &self,
+        relation_name: Option<&str>,
+        name: &str,
+    ) -> Result<DFField> {
+        if let Some(relation_name) = relation_name {
+            self.field_with_qualified_name(relation_name, name)
+        } else {
+            self.field_with_unqualified_name(name)
+        }
+    }
+
+    /// Find the field with the given name
+    pub fn field_with_unqualified_name(&self, name: &str) -> Result<DFField> {
+        let matches: Vec<&DFField> = self
+            .fields
+            .iter()
+            .filter(|field| field.name() == name)
+            .collect();
+        match matches.len() {
+            0 => Err(DataFusionError::Plan(format!("No field named '{}'", name))),
+            1 => Ok(matches[0].to_owned()),
+            _ => Err(DataFusionError::Plan(format!(
+                "Ambiguous reference to field named '{}'",
+                name
+            ))),
+        }
+    }
+
+    /// Find the field with the given qualified name
+    pub fn field_with_qualified_name(
+        &self,
+        relation_name: &str,
+        name: &str,
+    ) -> Result<DFField> {
+        let matches: Vec<&DFField> = self
+            .fields
+            .iter()
+            .filter(|field| {
+                field.qualifier == Some(relation_name.to_string()) && field.name() == name
+            })
+            .collect();
+        match matches.len() {
+            0 => Err(DataFusionError::Plan(format!(
+                "No field named '{}.{}'",
+                relation_name, name
+            ))),
+            1 => Ok(matches[0].to_owned()),
+            _ => Err(DataFusionError::Plan(format!(
```

Comment:
In theory this case should not be possible (as `new` would error out). Perhaps we should make this `unreachable!(..)` to signal it is not an expected usage error. Or maybe just change the error message to `"Internal Error: Ambiguous...."`)

Reply:
I changed the error type to `DataFusionError::Internal`